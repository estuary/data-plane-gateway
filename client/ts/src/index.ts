export * as protocols from "./protocols.ts";
export * as streams from "./streams.ts";
export { JournalClient } from "./journal_client.ts";
export { JournalSelector, ShardSelector } from "./selector.ts";
export { Result } from "./result.ts";
export { ShardClient } from "./shard_client.ts";
