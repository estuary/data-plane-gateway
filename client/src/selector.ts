import * as broker from "./gen/broker/protocol/broker.ts";
import { sortBy } from "./util.ts";

export class Selector {
  protected labels: Array<broker.ProtocolLabel>;

  constructor() {
    this.labels = [];
  }

  toLabelSet(): broker.ProtocolLabelSet {
    const sorted = this.labels.slice(0).sort(sortBy("name", "value"));
    return { labels: sorted };
  }

}

export class JournalSelector extends Selector {
  collection(v: string): this {
    this.labels.push({ name: "estuary.dev/collection", value: v });
    return this;
  }

  name(v: string): this {
    this.labels.push({ name: "name", value: v });
    return this;
  }

  static prefix(v: string): JournalSelector {
    if (!v.endsWith("/")) {
      v = v + "/";
    }
    const sel = new JournalSelector();
    sel.labels.push({ name: "prefix", value: v });
    return sel;
  }
}

export class ShardSelector extends Selector {
  task(v: string): this {
    this.labels.push({ name: "estuary.dev/task-name", value: v });
    return this;
  }

  id(v: string): this {
    this.labels.push({ name: "id", value: v });
    return this;
  }
}
