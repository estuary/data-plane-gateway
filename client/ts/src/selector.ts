import { broker } from "./protocols.ts";

export class Selector {
  protected labels: Array<broker.ProtocolLabel>;

  constructor() {
    this.labels = [];
  }

  toLabelSet(): broker.ProtocolLabelSet {
    return { labels: this.labels };
  }
}

export class JournalSelector extends Selector {
  collection(v: string): this {
    this.labels.push({ name: "estuary.dev/collection", value: v });
    return this;
  }

  name(v: string): this {
    this.labels.push({ name: "name", value: v });
    return this;
  }

  static prefix(v: string): JournalSelector {
    if (!v.endsWith("/")) {
      v = v + "/";
    }
    const sel = new JournalSelector();
    sel.labels.push({ name: "prefix", value: v });
    return sel;
  }
}

export class ShardSelector extends Selector {
  task(v: string): this {
    this.labels.push({ name: "estuary.dev/task-name", value: v });
    return this;
  }

  id(v: string): this {
    this.labels.push({ name: "id", value: v });
    return this;
  }
}
