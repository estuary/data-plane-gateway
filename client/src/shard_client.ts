import * as consumer from "./gen/consumer/protocol/consumer.ts";
import { Result } from "./result.ts";
import { ShardSelector } from "./selector.ts";
import { trimUrl } from "./util.ts";

export class ShardClient {
  private baseUrl: URL;
  private client: consumer.Api<null>;

  constructor(baseUrl: URL) {
    this.baseUrl = baseUrl;
    this.client = new consumer.Api({
      baseUrl: trimUrl(baseUrl),
    });
  }

  async list(
    include: ShardSelector = new ShardSelector(),
    exclude: ShardSelector = new ShardSelector(),
  ): Promise<Result<Array<consumer.ConsumerShardSpec>, Response>> {
    const response = await this.client.v1.shardList({
      selector: {
        include: include.toLabelSet(),
        exclude: exclude.toLabelSet(),
      },
    });

    if (response.ok) {
      return Result.Ok(response.data.shards!.map((j) => j.spec!));
    } else {
      return Result.Err(response);
    }
  }

  async stat(
    shard: string,
    readThrough: Record<string, string>,
  ): Promise<Result<consumer.ConsumerStatResponse, Response>> {
    const response = await this.client.v1.shardStat({ shard, readThrough });

    if (response.ok) {
      return Result.Ok(response.data);
    } else {
      return Result.Err(response);
    }
  }
}
