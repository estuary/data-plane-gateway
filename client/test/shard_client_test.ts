import { assertEquals, assertMatch } from "https://deno.land/std@0.135.0/testing/asserts.ts";
// TODO: Built-in snapshot testing has landed in deno std, but is not yet released.
import { test as snapshotTest } from "https://deno.land/x/snap/mod.ts";

import * as consumer from "../src/gen/consumer/protocol/consumer.ts";
import { BASE_URL, makeJwt } from "./test_support.ts";
import { ShardClient } from "../src/shard_client.ts";
import { ShardSelector } from "../src/selector.ts";


Deno.test("ShardClient.list task selector test", async () => {
  const client = new ShardClient(BASE_URL, await makeJwt({}));
  const taskSelector = new ShardSelector().task("acmeCo/source-hello-world");

  const shards = (await client.list(taskSelector)).unwrap();

  assertEquals(1, shards.length);
  assertEquals(
    "capture/acmeCo/source-hello-world/00000000-00000000",
    shards[0].spec.id,
  );
  assertEquals("PRIMARY", shards[0].status[0].code);
});

Deno.test("ShardClient.list bare id selector test", async () => {
  const client = new ShardClient(BASE_URL, await makeJwt({}));
  const idSelector = new ShardSelector().id(
    "capture/acmeCo/source-hello-world/00000000-00000000",
  );

  const error = (await client.list(idSelector)).unwrap_err();

  assertMatch(error.body.message!, new RegExp("No authorizing labels provided"));
});

Deno.test("ShardClient.list compound id selector test", async () => {
  const client = new ShardClient(BASE_URL, await makeJwt({}));
  const idSelector = new ShardSelector()
    .task("acmeCo/yet-another-task")
    .task("acmeCo/source-hello-world")
    .task("acmeCo/verifies-label-sorting")
    .id("capture/acmeCo/source-hello-world/00000000-00000000");

  const shards = (await client.list(idSelector)).unwrap();

  assertEquals(1, shards.length);
  assertEquals(
    "capture/acmeCo/source-hello-world/00000000-00000000",
    shards[0].spec.id,
  );
});

snapshotTest("ShardClient.stat test", async ({ assertSnapshot }) => {
  const client = new ShardClient(BASE_URL, await makeJwt({prefixes: ["capture/acmeCo/"]}));

  const stats = (await client.stat(
    "capture/acmeCo/source-hello-world/00000000-00000000",
    {},
  )).unwrap();

  const pluck = (res: consumer.ConsumerStatResponse) => {
    return {
      status: res.status,
      readThrough: res.readThrough,
      publishAt: res.publishAt,
    };
  };

  const masks = [
    "/readThrough/acmeCo\/source-hello-world\/eof",
    "/readThrough/acmeCo\/source-hello-world\/txn",
    "/publishAt/acmeCo\/greetings\/pivot=00",
    "/publishAt/ops\/acmeCo\/stats\/kind=capture\/name=acmeCo%2Fsource-hello-world\/pivot=00",
  ];
  assertSnapshot(pluck(stats), masks);
});
