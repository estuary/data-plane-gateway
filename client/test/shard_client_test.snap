{
  'ShardClient.list task selector test': [
    {
      spec: {
        id: 'capture/acmeCo/source-hello-world/00ffffffffffffff/00000000-00000000',
        sources: [],
        recoveryLogPrefix: 'recovery',
        hintPrefix: '/estuary/flow/hints',
        hintBackups: 2,
        maxTxnDuration: '1s',
        minTxnDuration: '0s',
        disable: false,
        hotStandbys: 0,
        labels: {
          labels: [
            {
              name: 'app.gazette.dev/managed-by',
              value: 'estuary.dev/flow',
            },
            {
              name: 'estuary.dev/build',
              value: '0000000000000000',
            },
            {
              name: 'estuary.dev/key-begin',
              value: '00000000',
            },
            {
              name: 'estuary.dev/key-end',
              value: 'ffffffff',
            },
            {
              name: 'estuary.dev/log-level',
              value: 'info',
            },
            {
              name: 'estuary.dev/rclock-begin',
              value: '00000000',
            },
            {
              name: 'estuary.dev/rclock-end',
              value: 'ffffffff',
            },
            {
              name: 'estuary.dev/task-name',
              value: 'acmeCo/source-hello-world',
            },
            {
              name: 'estuary.dev/task-type',
              value: 'capture',
            },
          ],
        },
        disableWaitForAck: false,
        ringBufferSize: 65536,
        readChannelSize: 4096,
      },
      status: [
        {
          code: 'FAILED',
          errors: [
            'runTransactions: txnStartCommit: store.StartCommit: failed to write atomic RocksDB commit\n\nCaused by:\n    IO error: No such file or directory: While open a file for appending: /home/travis/code/data-plane-gateway/test/tmp/capture_acmeCo_source-hello-world_00ffffffffffffff_00000000-00000000-1644699521/000008.log: No such file or directory',
          ],
        },
      ],
    },
  ],
  'ShardClient.stat test': {
    status: 'OK',
    readThrough: {
      'acmeCo/source-hello-world/txn': '[MASKED string]',
      'acmeCo/source-hello-world/eof': 'who cares',
    },
    publishAt: {
      'acmeCo/greetings/00ffffffffffffff/pivot=00': '[MASKED string]',
      'ops.us-central1.v1/stats/00ffffffffffffff/kind=capture/name=acmeCo%2Fsource-hello-world/pivot=00': '[MASKED string]',
    },
  },
}